include/ssm.h

```diff
@@ -663,6 +672,185 @@ ssm_value_t ssm_new_adt(uint8_t val_count, uint8_t tag);
+ *  @param arg      argument to be applied to @a closure.
+ *  @returns        a clone of @a closure with @a arg applied to it.
+ */
+ssm_value_t ssm_closure_apply_unsafe(ssm_value_t closure, ssm_value_t arg);
+
+/** @brief Partially apply a function to an argument.
+ *
+ *  This function strictly follows functional semantics: it leaves @a closure
+ *  unmodified, and returns a copy of @a closure with the additional @a arg
+ *  argument. ssm_dup() is called on all arguments of @a closure, as well as the
+ *  new argument @a arg.
+ *
+ *  It is the responsibility of the caller to ensure that @a closure has enough
+ *  capacity to accommodate @a arg. The result of calling this when it no longer
+ *  has capacity is undefined.
+ *
+ *  @param closure  value pointing to a heap-allocated closure.
+ *  @param arg      argument to be applied to @a closure.
+ *  @returns        a clone of @a closure with @a arg applied to it.
+ */
+#define ssm_closure_apply(closure, arg)                                        \
+  ssm_closure_apply_unsafe(closure, ssm_dup(arg))
+
+/** @brief Invoke a closure's enter function and activate the new process.
+ *
+ *  Note that @a closure does not have to have its @a arg_count set to capacity,
+ *  as long as all slots in @a argv are written to. This means that this macro
+ *  is safe to invoke after an argument is added with
+ *  ssm_closure_store_arg_unsafe(), without needing to increment @a arg_count.
+ *
+ *  This macro is unsafe in that it does not perform any memory management for
+ *  the closure, e.g., calling ssm_dup() on its arguments.
+ *
+ *  @param closure  value pointing to a heap-allocated closure.
+ *  @param parent   the activation record of the caller.
+ *  @param prio     the priority of the callee.
+ *  @param depth    the depth of the callee.
+ *  @param ret      the return value address that the callee should write to.
+ */
+#define ssm_closure_activate_unsafe(closure, parent, prio, depth, ret)         \
+  do {                                                                         \
+    ssm_activate(ssm_closure_func(closure)(parent, prio, depth,                \
+                                           ssm_closure_argv(closure), ret));   \
+  } while (0)
+
+/** @brief Apply a closure to its final argument, and activate the process.
+ *
+ *  This macro assumes that @a closure has all but one argument aleady applied.
+ *  It will call ssm_dup() on all values in @a argv, since they will be passed
+ *  to the newly created process.
+ *
+ *  @param closure  value pointing to a heap-allocated closure.
+ *  @param arg      final argument applied to @a closure.
+ *  @param parent   the activation record of the caller.
+ *  @param prio     the priority of the callee.
+ *  @param depth    the depth of the callee.
+ *  @param ret      the return value address that the callee should write to.
+ */
+#define ssm_closure_reduce(closure, arg, parent, prio, depth, ret)             \
```

Comment:
Reduce is just a macro now. It doesn't even allocate a new closure; it just temporarily writes `arg` to the `argv` of `closure`, and calls the enter function. We don't even need to bother taking out the `arg`; because we don't increment `arg_count` (i.e., the `val_count` field of `mm`), it is never "owned" by the closure.

Reply:
I removed this altogether